src/cli/commands/add.js

```diff
@@ -55,23 +56,22 @@ export class Add extends Install {
    */
   getPatternVersion(pattern: string, pkg: Manifest): string {
     const {exact, tilde} = this.flags;
-    const parts = PackageRequest.normalizePattern(pattern);
+    const {hasVersion, range} = PackageRequest.normalizePattern(pattern);
     let version;
+
     if (getExoticResolver(pattern)) {
       // wasn't a name/range tuple so this is just a raw exotic pattern
       version = pattern;
-    } else if (parts.hasVersion && parts.range) {
+    } else if (hasVersion && range && (semver.satisfies(pkg.version, range) || getExoticResolver(range))) {
       // if the user specified a range then use it verbatim
-      version = parts.range === 'latest' ? `^${pkg.version}` : parts.range;
-    } else if (tilde) {
-      // --save-tilde
-      version = `~${pkg.version}`;
-    } else if (exact) {
-      // --save-exact
-      version = pkg.version;
+      version = range;
```

Comment:
It really shouldn't since `semver.satisfies` should return false when `range` is `"latest"`. I verified this locally and tests pass locally. I'll see what's wrong there and fix it.